src/store.rs

```diff
@@ -8,130 +9,159 @@ use zkevm_opcode_defs::{
-        }
-    }
+#[derive(Debug)]
+pub struct StateStorage {
+    pub storage_changes: HashMap<StorageKey, U256>,
+    pub initial_storage: Rc<RefCell<dyn InitialStorage>>,
+    l1_to_l2_logs: Vec<L2ToL1Log>,
+}
 
-    pub fn new(
-        contract_storage: HashMap<U256, Vec<U256>>,
-        state_storage: HashMap<StorageKey, U256>,
-    ) -> Self {
+impl Default for StateStorage {
+    fn default() -> Self {
         Self {
-            contract_storage,
-            state_storage,
+            storage_changes: HashMap::new(),
+            initial_storage: Rc::new(RefCell::new(InitialStorageMemory {
+                initial_storage: HashMap::new(),
+            })),
+            l1_to_l2_logs: Vec::new(),
         }
     }
 }
 
-impl Storage for InMemory {
-    fn decommit(&self, hash: U256) -> Result<Option<Vec<U256>>, StorageError> {
-        Ok(self.contract_storage.get(&hash).cloned())
-    }
-
-    fn add_contract(&mut self, hash: U256, code: Vec<U256>) -> Result<(), StorageError> {
-        self.contract_storage.insert(hash, code);
-        Ok(())
+impl StateStorage {
+    pub fn new(initial_storage: Rc<RefCell<dyn InitialStorage>>) -> Self {
+        Self {
+            storage_changes: HashMap::new(),
+            initial_storage,
+            l1_to_l2_logs: Vec::new(),
+        }
     }
 
-    fn storage_read(&self, key: StorageKey) -> Result<Option<U256>, StorageError> {
-        Ok(self.state_storage.get(&key).copied())
+    pub fn storage_read(&self, key: StorageKey) -> Result<Option<U256>, StorageError> {
+        let mut value = self.storage_changes.get(&key).copied();
+        if value.is_none() {
+            value = self.initial_storage.borrow().storage_read(key)?;
+        }
+        Ok(value)
     }
 
-    fn storage_write(&mut self, key: StorageKey, value: U256) -> Result<(), StorageError> {
-        self.state_storage.insert(key, value);
+    pub fn storage_write(&mut self, key: StorageKey, value: U256) -> Result<(), StorageError> {
+        self.storage_changes.insert(key, value);
         Ok(())
     }
 
-    fn storage_drop(&mut self, key: StorageKey) -> Result<(), StorageError> {
-        self.state_storage.remove(&key);
+    pub fn record_l2_to_l1_log(&mut self, msg: L2ToL1Log) -> Result<(), StorageError> {
+        self.l1_to_l2_logs.push(msg);
         Ok(())
     }
 
-    fn get_state_storage(&self) -> &HashMap<StorageKey, U256> {
-        &self.state_storage
+    pub fn create_snapshot(&self) -> SnapShot {
+        SnapShot {
+            storage_changes: self.storage_changes.clone(),
+        }
     }
 
-    fn get_all_keys(&self) -> Vec<StorageKey> {
-        self.state_storage.keys().copied().collect()
+    pub fn rollback(&mut self, snapshot: &SnapShot) {
+        let keys = snapshot.storage_changes.keys();
+        for key in keys {
+            let value = snapshot.storage_changes.get(key);
+            if let Some(value) = value {
+                self.storage_write(*key, *value).unwrap();
+            }
```

Comment:
Another (optional) style suggestion:
```suggestion
                 snapshot
                .storage_changes
                .get(key)
                .map(|value| self.storage_write(*key, *value));
```